doc/base64tool.1
.TH BASE64TOOL "1" "July 2009" "base64tool v2.9.7" "User Commands"


.SH NAME
base64tool \- encode files into Base64 encoding and decode Base64 encoded files


.SH SYNOPSIS
.B base64tool
<\fIencode\fR|\fIdecode\fR> <\fIinput\fR> <\fIoutput\fR> [\fImaxlen\fR]


.SH DESCRIPTION
.LP
This program can either encode a file with Base64 encoding suited for
inclusion of binary data into text files such as XML files and decode
Base64 encoded data back into its original form. The first argument
controls whether \fBbase64tool\fR will encode or decode. The second
argument is name of the file that data is read from. The third
argument is the name of the file that output is written to. The file
will be overwritten if it exists.
.LP
When encoding an optional fourth parameter, \fImaxlen\fR, specifies
the maximum line length to output. It defaults to 72 if left out.


.SH AUTHOR
.I base64tool
was written by Moritz Bunkus <moritz@bunkus.org>.
.SH WWW
The newest version can always be found at
.UR http://www.bunkus.org/\:videotools/\:mkvtoolnix/
.UE


ChangeLog
2009-07-06  Moritz Bunkus  <moritz@bunkus.org>

	* mkvmerge: bug fix: Fixed the audio sync for tracks read from AVI
	files containing garbage at the beginning. Fix for bug 421.

	* mmg: bug fix: Trying to save chapters that contain editions
	without a single chapter entry does no longer result in a crash
	but a descriptive error message instead. Saving empty chapters to
	a Matroska file will remove all chapters contained in the file
	instead of not doing anything. Fixes for bug 422.

2009-07-05  Moritz Bunkus  <moritz@bunkus.org>

	* mkvinfo: enhancement: Implemented speed-ups of up to 50% for
	middle to larger files.

2009-07-02  Moritz Bunkus  <moritz@bunkus.org>

	* mkvmerge: bug fix: Fixed reading AVC/h.264 video tracks from OGM
	files. Fix for bug 418.

	* mmg: bug fix: The chapter language for chapters copied from
	source files (e.g. Matroska, MP4 or OGM files) is only changed if
	the user has selected any language other than "und". Fix for bug
	420.

	* mmg: bug fix: mmg will no longer show an error message if the
	user has not selected a country in the "chapters" tab of the
	preferences dialog.

2009-07-01  Moritz Bunkus  <moritz@bunkus.org>

	* Released v2.9.7.

2009-06-28  Moritz Bunkus  <moritz@bunkus.org>

	* mkvmerge: bug fix: The handling of NVOPs in native MPEG4 part 2
	video storage has been improved. NVOPs are dropped again both from
	packed and non-packed bitstreams, and timecodes are adjusted to
	match the number of dropped frames.

2009-06-25  Moritz Bunkus  <moritz@bunkus.org>

	* mmg: The "preferences" dialog has been split up into several
	tabs. Some other preferences available in other dialogs have been
	merged into the "preferences" dialog.

	* mkvmerge: bug fix: The I frame detection for AVC/h.264 video has
	been fixed. Sometimes a single I frame was recognized as two or
	more consecutive I frames resulting in garbled display and wrong
	timestamps. Fix for bug 415.

2009-06-23  Moritz Bunkus  <moritz@bunkus.org>

	* mmg: new feature: The list of common languages can be configured
	by the user via the 'preferences' dialog.

2009-06-22  Moritz Bunkus  <moritz@bunkus.org>

	* all: bug fix: The programs do not try to close iconv handles -1
	anymore which resulted in segfault during uninitialization on some
	platforms (e.g. FreeBSD, Mac OS X). Fix for bug 412.

2009-06-21  Moritz Bunkus  <moritz@bunkus.org>

	* mkvmerge: bug fix: Complete rewrite of the code for the native
	storage mode for MPEG4 part 2 video tracks. Fix for bug 298.

2009-06-18  Moritz Bunkus  <moritz@bunkus.org>

	* mkvmerge: bug fix: Made the detection rules for raw MP3, AC3 and
	AAC audio files more strict. This avoids a mis-detection of
	certain files, e.g. AVC/h.264 ES files being misdetected as MP3
	files. Fix for bug 414.

2009-06-14  Moritz Bunkus  <moritz@bunkus.org>

	* mkvmerge: bug fix: Appending MP4 or OGM files with chapters will
	merge the chapters from all appended files and not just take the
	chapters from the first file and discard the chapters from the
	following files. Fix for bug 397.

	* mkvmerge, mmg: new feature: The language for chapters read from
	files such as OGM and MP4 files can be chosen with the
	--chapter-language command line option. Fix for bug 399.

	* mkvmerge, mmg: new feature: Chapter and tag information
	contained in source files are now shown in the "track" selection
	box and can be toggled individually. The user can set the charset
	for chapters if the source file's chapters are not encoded in
	UTF-8 (e.g. in some OGM/MP4 files). The old file specific
	checkboxes "no tags" and "no chapters" have been removed. Fix for
	bug 400.

2009-06-13  Moritz Bunkus  <moritz@bunkus.org>

	* mmg: bug fix: The chapter editor was corrupting Matroska files
	if the chapters were saved to a file twice in a row without
	reloading them from the Matroska file.

2009-06-09  Moritz Bunkus  <moritz@bunkus.org>

	* mmg: enhancement: If the user selects the option "Verify" from
	the "Chapters" menu then a message will be shown even if the
	validation succeeded. Fix for bug 410.

	* mmg/mkvmerge: bug fix: Adding MP4 files with chapter entries
	that are not encoded in UTF-8 will not result in mkvmerge aborting
	with a message about invalid UTF-8 sequences anymore. Fix for bug
	408.

2009-06-07  Moritz Bunkus  <moritz@bunkus.org>

	* Released v2.9.5.

2009-06-06  Moritz Bunkus  <moritz@bunkus.org>

	* mkvmerge: bug fix: The handling of TrueHD/MLP audio in MPEG
	program streams was broken resulting in many dropped packets.

	* all: bug fix: There was the possibility that invalid memory
	access occured and e.g. mkvmerge crashed on systems with the
	posix_fadvise() function (mainly Linux) if mkvmerge was opening
	several files from certain file systems (e.g. VFAT). Apart from
	obvious crashes the only other side effect was that the
	posix_fadvise() function would not be used resulting in slightly
	worse I/O performance.

2009-05-31  Moritz Bunkus  <moritz@bunkus.org>

	* mkvmerge: new feature: Improved the control over which tags get
	copied from a source file to the output file. The old option
	"--no-tags" was replaced with the new options "--no-global-tags"
	which causes global tags not to be copied and "--no-track-tags"
	which causes track specific tags to not be copied. The new option
	"--track-tags" can be used to select tracks for which tags will be
	copied. The default is still to copy all existing tags.

	* mkvmerge: new feature: Included chapters, global and track
	specific tracks in the output of mkvmerge's identification mode.

	* mkvmerge: Renamed a couple of command line options to make the
	command line interface more consistent: "--no-audio",
	"--no-video", "--no-subtitles", "--no-buttons", "--audio-tracks",
	"--video-tracks", "--subtitle-tracks", "--button-tracks". The old
	versions of these options "--noaudio", "--novideo", "--nosubs",
	"--nobuttons", "--atracks", "--vtracks", "--stracks" and
	"--btracks" still work.


src/tools/base64tool.cpp
/*
   base64util - Utility for encoding and decoding Base64 files.

   Distributed under the GPL
   see the file COPYING for details
   or visit http://www.gnu.org/copyleft/gpl.html

   command line parameter parsing, looping, output handling

   Written by Moritz Bunkus <moritz@bunkus.org>.
*/

#include "common/os.h"

#include <errno.h>
#include <string.h>

#include <string>

#include "common/base64.h"
#include "common/command_line.h"
#include "common/common.h"
#include "common/mm_io.h"
#include "common/strings/editing.h"
#include "common/strings/parsing.h"


void
set_usage() {
  usage_text = Y(
    "base64util <encode|decode> <input> <output> [maxlen]\n"
    "\n"
    "  encode - Read from <input>, encode to Base64 and write to <output>.\n"
    "           Max line length can be specified and is 72 if left out.\n"
    "  decode - Read from <input>, decode to binary and write to <output>.\n"
    );

  version_info = "base64util v" VERSION;
}

int
main(int argc,
     char *argv[]) {
  int maxlen;
  uint64_t size;
  unsigned char *buffer;
  char mode;
  std::string s, line;
  mm_io_c *in, *out;
  mm_text_io_c *intext;

  init_stdio();
  set_usage();

  if (argc < 4)
    usage(0);

  mode = 0;
  in = NULL;
  out = NULL;
  intext = NULL;
  if (!strcmp(argv[1], "encode"))
    mode = 'e';
  else if (!strcmp(argv[1], "decode"))
    mode = 'd';
  else
    mxerror(boost::format(Y("Invalid mode '%1%'.\n")) % argv[1]);

  maxlen = 72;
  if ((argc == 5) && (mode == 'e')) {
    if (!parse_int(argv[4], maxlen) || (maxlen < 4))
      mxerror(Y("Max line length must be >= 4.\n\n"));
  } else if ((argc > 5) || ((argc > 4) && (mode == 'd')))
    usage(2);

  maxlen = ((maxlen + 3) / 4) * 4;

  try {
    in = new mm_file_io_c(argv[2]);
    if (mode != 'e')
      intext = new mm_text_io_c(in);
  } catch(...) {
    mxerror(boost::format(Y("The file '%1%' could not be opened for reading (%2%, %3%).\n")) % argv[2] % errno % strerror(errno));
  }

  try {
    out = new mm_file_io_c(argv[3], MODE_CREATE);
  } catch(...) {
    mxerror(boost::format(Y("The file '%1%' could not be opened for writing (%2%, %3%).\n")) % argv[3] % errno % strerror(errno));
  }

  in->save_pos();
  in->setFilePointer(0, seek_end);
  size = in->getFilePointer();
  in->restore_pos();

  if (mode == 'e') {
    buffer = (unsigned char *)safemalloc(size);
    size = in->read(buffer, size);
    delete in;

    s = base64_encode(buffer, size, true, maxlen);
    safefree(buffer);

    out->write(s.c_str(), s.length());
    delete out;

  } else {

    while (intext->getline2(line)) {
      strip(line);
      s += line;
    }
    delete intext;

    buffer = (unsigned char *)safemalloc(s.length() / 4 * 3 + 100);
    try {
      size = base64_decode(s, buffer);
    } catch(...) {
      delete in;
      delete out;
      mxerror(Y("The Base64 encoded data could not be decoded.\n"));
    }
    out->write(buffer, size);

    safefree(buffer);
    delete out;
  }

  mxinfo(Y("Done.\n"));

  return 0;
}
